### DISCUSSION TITLE
Cascade update generates n+1 queries

### DISCUSSION THREAD

**Discussion by Michael881188** [nhibernate/nhibernate-core#3042] [General]
For example, I have:
```
public UserMap() {
  Id(x => x.Id);
  HasMany(x => x.Attributes).Cascade.AllDeleteOrphan();
}

public AttributeMap() {
  Id(x => x.Id);
  References(x => x.User).Not.Nullable();
}

public class User {
  public virtual Guid Id { get; set; }
  public virtual ISet<Attribute> Attributes { get; protected set; } =
            new HashSet<Attribute>();
}

public class Attribute {
  public virtual Guid Id { get; set; }
  public virtual User User{ get; protected set; }
}
```
And now, whenever I'm adding anything into Attributes array in User entity and saving it(collection of attributes is fully fetched before that). It generates n+1 queries. It is selecting 1 by 1 each Attribute and after that inserting all of them in 1 query. For ex. 10 attributes generate: 10 selects + 1 insert query.
I'm not sure, maybe something is wrong with my mapping but I didn't find any solution.

**Comment by hazzik** [comment]
It is not clear what is the shape of your entities. As I can see you have unidirectional entities? The likely cause of SELECT N+1 is updatable keys on relationships. Please refer to this SO answer: https://stackoverflow.com/a/7601312/259946 to fix the mapping.

**Comment by Michael881188** [comment]
I have bidirectional entities. I updated the mapping and added example classes.
